File: 백준/Gold/23254. 나는 기말고사형 인간이야/나는 기말고사형 인간이야.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        StringTokenizer aSt = new StringTokenizer(br.readLine());
        StringTokenizer bSt = new StringTokenizer(br.readLine());

        PriorityQueue<Grade> pq = new PriorityQueue<>();
        for (int i = 0; i < m; i++) {
            pq.add(new Grade(Integer.parseInt(aSt.nextToken()), Integer.parseInt(bSt.nextToken())));
        }
        int time = n * 24;
        int result = 0;
        while (time > 0 && !pq.isEmpty()) {

            Grade grade = pq.poll();
            if (grade.current == 100) {
                result += 100;
                continue;
            }
            int useTime = Math.min(grade.calculate(), time);
            time -= useTime;
            pq.add(grade.study(useTime));
        }
        while (!pq.isEmpty()) {
            result += pq.poll().current;
        }
        System.out.println(result);
    }

    private static class Grade implements Comparable<Grade> {
        int current;
        int performance;

        public Grade(int current, int performance) {
            this.current = current;
            this.performance = performance;
        }

        public int calculate() {
            return (100 - current) / performance;
        }

        public Grade study(final int time) {
            int next = current + performance * time;
            return new Grade(next, 100 - next);
        }

        @Override
        public int compareTo(Grade o) {
            return Integer.compare(o.performance, performance);
        }
    }
}

File: 백준/Gold/23254. 나는 기말고사형 인간이야/나는 기말고사형 인간이야.kt
import java.util.PriorityQueue
import java.util.StringTokenizer
import kotlin.math.min

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }

    val aSt = StringTokenizer(readLine())
    val bSt = StringTokenizer(readLine())

    val pq = PriorityQueue<Grade> { o1, o2 -> o2.performance.compareTo(o1.performance) }
    repeat(m) {
        val current = aSt.nextToken().toInt()
        val performance = bSt.nextToken().toInt()
        pq.add(Grade(current, performance))
    }

    var time = n * 24
    var result = 0

    while (time > 0 && pq.isNotEmpty()) {
        val grade = pq.poll()
        if (grade.current == 100) {
            result += 100
            continue
        }
        val useTime = min(grade.calculate(), time)
        time -= useTime
        pq.add(grade.study(useTime))
    }

    while (pq.isNotEmpty()) {
        result += pq.poll().current
    }

    println(result)
}

data class Grade(
    val current: Int,
    val performance: Int
) {
    fun calculate(): Int {
        return (100 - current) / performance
    }

    fun study(time: Int): Grade {
        val next = current + performance * time
        return Grade(next, 100 - next)
    }
}
